### PR TITLE
[TE] detection - moving monitoring window aligned to data set granularity

### PR DESCRIPTION
**Pull request by jihaozh**
- moving monitoring window 
- moving window aligned to dataset granularity

### CODE REVIEW

File: thirdeye/thirdeye-pinot/src/main/java/com/linkedin/thirdeye/detection/algorithm/stage/AnomalyDetectionStageWrapper.java [hunk 1]
```diff
@@ -17,46 +17,98 @@
...
+
   private final String metricUrn;
+  private final Map<String, Object> specs;
+  private final String stageClassName;
+
+  private final int windowDelay;
+  private final TimeUnit windowDelayUnit;
+  private final int windowSize;
+  private final TimeUnit windowUnit;
+  private final boolean movingWindowDetection;
```
Comment (akshayrai) [inline thirdeye/thirdeye-pinot/src/main/java/com/linkedin/thirdeye/detection/algorithm/stage/AnomalyDetectionStageWrapper.java]: I'd prefer naming this as isMovingWindowDetection

File: thirdeye/thirdeye-pinot/src/main/java/com/linkedin/thirdeye/detection/algorithm/stage/AnomalyDetectionStageWrapper.java [hunk 2]
```diff
@@ -17,46 +17,98 @@
...
     this.metricUrn = MapUtils.getString(config.getProperties(), PROP_METRIC_URN);
-    if (this.metricUrn != null){
-      specs.put(PROP_METRIC_URN, metricUrn);
+    if (this.metricUrn != null) {
+      this.specs.put(PROP_METRIC_URN, metricUrn);
     }
-    this.anomalyDetectionStage.init(specs, config.getId(), startTime, endTime);
+
+    this.movingWindowDetection = MapUtils.getBooleanValue(this.specs, PROP_MOVING_WINDOW_DETECTION, false);
+    this.windowDelay = MapUtils.getIntValue(this.specs, PROP_WINDOW_DELAY, 0);
```
Comment (akshayrai) [inline thirdeye/thirdeye-pinot/src/main/java/com/linkedin/thirdeye/detection/algorithm/stage/AnomalyDetectionStageWrapper.java]: Can you include some comments in the code on what and why we need this windowDelay for?

File: thirdeye/thirdeye-pinot/src/main/java/com/linkedin/thirdeye/detection/algorithm/stage/AnomalyDetectionStageWrapper.java [hunk 3]
```diff
@@ -67,4 +119,86 @@ public DetectionPipelineResult run() {
...
+
+  private List<Long> getMonitoringWindowEndTimes() {
+    List<Long> endTimes = new ArrayList<>();
+
+    // get current hour/day, depending on granularity of dataset,
+    DateTime currentEndTime = new DateTime(getBoundaryAlignedTimeForDataset(new DateTime(endTime, dateTimeZone)), dateTimeZone);
+
+    DateTime lastDateTime = new DateTime(getBoundaryAlignedTimeForDataset(new DateTime(startTime, dateTimeZone)), dateTimeZone);
+    Period bucketSizePeriod = getBucketSizePeriodForDataset();
+    while (lastDateTime.isBefore(currentEndTime)) {
```
Comment (akshayrai) [inline thirdeye/thirdeye-pinot/src/main/java/com/linkedin/thirdeye/detection/algorithm/stage/AnomalyDetectionStageWrapper.java]: isBefore does a less than check right? Won't currentEndTime be excluded then?

File: thirdeye/thirdeye-pinot/src/main/java/com/linkedin/thirdeye/detection/algorithm/stage/AnomalyDetectionStageWrapper.java [hunk 4]
```diff
@@ -67,4 +119,86 @@ public DetectionPipelineResult run() {
   private AnomalyDetectionStage loadAnomalyDetectorStage(String className) throws Exception {
     return (AnomalyDetectionStage) Class.forName(className).newInstance();
   }
+
+  List<Interval> getMonitoringWindows() {
```
Comment (akshayrai) [inline thirdeye/thirdeye-pinot/src/main/java/com/linkedin/thirdeye/detection/algorithm/stage/AnomalyDetectionStageWrapper.java]: I'd prefer passing startTime and endTime as parameters. We can convert this into a utility in the future.

File: thirdeye/thirdeye-pinot/src/main/java/com/linkedin/thirdeye/detection/algorithm/stage/AnomalyDetectionStageWrapper.java (hunk 1 above)
Comment (jihaozh) [reply thirdeye/thirdeye-pinot/src/main/java/com/linkedin/thirdeye/detection/algorithm/stage/AnomalyDetectionStageWrapper.java]: changed to isMovingWindowDetection

File: thirdeye/thirdeye-pinot/src/main/java/com/linkedin/thirdeye/detection/algorithm/stage/AnomalyDetectionStageWrapper.java (hunk 2 above)
Comment (jihaozh) [reply thirdeye/thirdeye-pinot/src/main/java/com/linkedin/thirdeye/detection/algorithm/stage/AnomalyDetectionStageWrapper.java]: added

File: thirdeye/thirdeye-pinot/src/main/java/com/linkedin/thirdeye/detection/algorithm/stage/AnomalyDetectionStageWrapper.java (hunk 3 above)
Comment (jihaozh) [reply thirdeye/thirdeye-pinot/src/main/java/com/linkedin/thirdeye/detection/algorithm/stage/AnomalyDetectionStageWrapper.java]: yes. this is the expected behavior as we do currently.

File: thirdeye/thirdeye-pinot/src/main/java/com/linkedin/thirdeye/detection/algorithm/stage/AnomalyDetectionStageWrapper.java (hunk 4 above)
Comment (jihaozh) [reply thirdeye/thirdeye-pinot/src/main/java/com/linkedin/thirdeye/detection/algorithm/stage/AnomalyDetectionStageWrapper.java]: maybe not now cause this function requires passing multiple parameters from outside and intended to use only here.

**Pull request merged**